File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/grpc/consumer/ConsumerManager.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.eventmesh.runtime.core.protocol.grpc.consumer;

import lombok.extern.slf4j.Slf4j;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.eventmesh.common.protocol.SubscriptionMode;
import org.apache.eventmesh.common.utils.JsonUtils;
import org.apache.eventmesh.common.utils.LogUtils;
import org.apache.eventmesh.runtime.boot.EventMeshGrpcServer;
import org.apache.eventmesh.runtime.common.ServiceState;
import org.apache.eventmesh.runtime.core.protocol.grpc.consumer.consumergroup.ConsumerGroupClient;
import org.apache.eventmesh.runtime.core.protocol.grpc.consumer.consumergroup.GrpcType;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

@Slf4j
public class ConsumerManager {

    private final EventMeshGrpcServer eventMeshGrpcServer;

    private final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();

    // key: ConsumerGroup
    private final Map<String, List<ConsumerGroupClient>> clientTable = new ConcurrentHashMap<>();

    // key: ConsumerGroup
    private final Map<String, EventMeshConsumer> consumerTable = new ConcurrentHashMap<>();

    public ConsumerManager(final EventMeshGrpcServer eventMeshGrpcServer) {
        this.eventMeshGrpcServer = eventMeshGrpcServer;
    }

    public Map<String, List<ConsumerGroupClient>> getClientTable() {
        return clientTable;
    }

    public void init() throws Exception {
        LogUtils.info(log, "Grpc ConsumerManager initialized.");
    }

    public void start() throws Exception {
        startClientCheck();
        LogUtils.info(log, "Grpc ConsumerManager started.");
    }

    public void shutdown() throws Exception {
        for (final EventMeshConsumer consumer : consumerTable.values()) {
            consumer.shutdown();
        }
        scheduledExecutorService.shutdown();
        LogUtils.info(log, "Grpc ConsumerManager shutdown.");
    }

    public EventMeshConsumer getEventMeshConsumer(final String consumerGroup) {
        return consumerTable.computeIfAbsent(consumerGroup, key -> new EventMeshConsumer(eventMeshGrpcServer, consumerGroup));
    }

    public synchronized void registerClient(final ConsumerGroupClient newClient) {
        final String consumerGroup = newClient.getConsumerGroup();
        final String topic = newClient.getTopic();
        final GrpcType grpcType = newClient.getGrpcType();
        final String url = newClient.getUrl();
        final String ip = newClient.getIp();
        final String pid = newClient.getPid();
        final SubscriptionMode subscriptionMode = newClient.getSubscriptionMode();

        List<ConsumerGroupClient> localClients = clientTable.get(consumerGroup);
        if (localClients == null) {
            localClients = new ArrayList<>();
            localClients.add(newClient);
            clientTable.putIfAbsent(consumerGroup, localClients);
        } else {
            boolean isContains = false;
            for (final ConsumerGroupClient localClient : localClients) {
                if (GrpcType.WEBHOOK == grpcType && StringUtils.equals(localClient.getTopic(), topic)
                        && StringUtils.equals(localClient.getUrl(), url)
                        && localClient.getSubscriptionMode() == subscriptionMode) {
                    isContains = true;
                    localClient.setUrl(newClient.getUrl());
                    localClient.setLastUpTime(newClient.getLastUpTime());
                    break;
                } else if (GrpcType.STREAM == grpcType && StringUtils.equals(localClient.getTopic(), topic)
                        && StringUtils.equals(localClient.getIp(), ip) && StringUtils.equals(localClient.getPid(), pid)
                        && localClient.getSubscriptionMode() == subscriptionMode) {
                    isContains = true;
                    localClient.setEventEmitter(newClient.getEventEmitter());
                    localClient.setLastUpTime(newClient.getLastUpTime());
                    break;
                }
            }
            if (!isContains) {
                localClients.add(newClient);
            }
        }
    }

    public boolean updateClientTime(final ConsumerGroupClient client) {
        final List<ConsumerGroupClient> localClients = clientTable.get(client.getConsumerGroup());
        if (CollectionUtils.isEmpty(localClients)) {
            return false;
        }

        for (final ConsumerGroupClient localClient : localClients) {
            if (StringUtils.equals(localClient.getIp(), client.getIp())
                    && StringUtils.equals(localClient.getPid(), client.getPid())
                    && StringUtils.equals(localClient.getSys(), client.getSys())
                    && StringUtils.equals(localClient.getTopic(), client.getTopic())) {
                localClient.setLastUpTime(new Date());
                return true;
            }
        }

        return false;
    }

    public synchronized void deregisterClient(final ConsumerGroupClient client) {
        final String consumerGroup = client.getConsumerGroup();
        final List<ConsumerGroupClient> localClients = clientTable.get(consumerGroup);
        if (CollectionUtils.isEmpty(localClients)) {
            return;
        }

        final Iterator<ConsumerGroupClient> iterator = localClients.iterator();
        synchronized (clientTable) {
            while (iterator.hasNext()) {
                final ConsumerGroupClient localClient = iterator.next();
                if (StringUtils.equals(localClient.getTopic(), client.getTopic())
                        && localClient.getSubscriptionMode() == client.getSubscriptionMode()) {
                    // close the GRPC client stream before removing it
                    closeEventStream(localClient);
                    iterator.remove();
                }
            }
        }

        if (CollectionUtils.isEmpty(localClients)) {
            clientTable.remove(consumerGroup);
        }

    }

    private void closeEventStream(final ConsumerGroupClient client) {
        if (client.getEventEmitter() != null) {
            client.getEventEmitter().onCompleted();
        }
    }

    public synchronized void restartEventMeshConsumer(final String consumerGroup) throws Exception {
        final EventMeshConsumer eventMeshConsumer = consumerTable.get(consumerGroup);

        if (eventMeshConsumer == null) {
            return;
        }

        if (ServiceState.RUNNING == eventMeshConsumer.getStatus()) {
            eventMeshConsumer.shutdown();
        }

        eventMeshConsumer.init();
        eventMeshConsumer.start();

        if (ServiceState.RUNNING != eventMeshConsumer.getStatus()) {
            consumerTable.remove(consumerGroup);
        }
    }

    private void startClientCheck() {
        final int clientTimeout = eventMeshGrpcServer.getEventMeshGrpcConfiguration().getEventMeshSessionExpiredInMills();
        if (clientTimeout > 0) {
            scheduledExecutorService.scheduleAtFixedRate(() -> {
                LogUtils.debug(log, "grpc client info check");

                final List<ConsumerGroupClient> clientList = new ArrayList<>();
                clientTable.values().forEach(clientList::addAll);

                LogUtils.debug(log, "total number of ConsumerGroupClients: {}", clientList.size());

                if (CollectionUtils.isEmpty(clientList)) {
                    return;
                }

                final Set<String> consumerGroupRestart = new HashSet<>();
                clientList.forEach(client -> {
                    if (System.currentTimeMillis() - client.getLastUpTime().getTime() > clientTimeout) {
                        LogUtils.warn(log, "client {} lastUpdate time {} over three heartbeat cycles. Removing it",
                                JsonUtils.toJSONString(client), client.getLastUpTime());

                        deregisterClient(client);
                        if (getEventMeshConsumer(client.getConsumerGroup()).deregisterClient(client)) {
                            consumerGroupRestart.add(client.getConsumerGroup());
                        }
                    }
                });

                // restart EventMeshConsumer for the group
                consumerGroupRestart.forEach(consumerGroup -> {
                    try {
                        restartEventMeshConsumer(consumerGroup);
                    } catch (Exception e) {
                        LogUtils.error(log, "Error in restarting EventMeshConsumer [{}]", consumerGroup, e);
                    }
                });
            }, 10_000, 10_000, TimeUnit.MILLISECONDS);
        }
    }

    public List<String> getAllConsumerTopic() {
        return clientTable.values()
                .stream()
                .flatMap(List::stream)
                .map(ConsumerGroupClient::getTopic)
                .distinct()
                .collect(Collectors.toList());
    }
}

File: eventmesh-server-go/runtime/core/protocol/grpc/consumer/consumer_manager.go
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package consumer

import (
	config2 "github.com/apache/incubator-eventmesh/eventmesh-server-go/config"
	"github.com/apache/incubator-eventmesh/eventmesh-server-go/log"
	"github.com/apache/incubator-eventmesh/eventmesh-server-go/runtime/consts"
	"github.com/liyue201/gostl/ds/set"
	"github.com/pkg/errors"
	"sync"
	"time"
)

var (
	ErrNoConsumerClient = errors.New("no consumer group client")
)

//go:generate mockgen -destination ./mocks/consumer_manager.go -package mocks github.com/apache/incubator-eventmesh/eventmesh-server-go/runtime/core/protocol/grpc/consumer ConsumerManager
type ConsumerManager interface {
	GetConsumer(consumerGroup string) (EventMeshConsumer, error)
	RegisterClient(cli *GroupClient) error
	DeRegisterClient(cli *GroupClient) error
	UpdateClientTime(cli *GroupClient)
	RestartConsumer(consumerGroup string) error
	Start() error
	Stop() error
}

type consumerManager struct {
	// consumerClients store all consumer clients
	// key is consumer group, value is set *GroupClient
	consumerGroupClients *sync.Map

	// consumers eventmesh consumer instances
	// key is consumer group, value is EventMeshConsumer
	consumers *sync.Map
}

// NewConsumerManager create new consumer manager
func NewConsumerManager() (ConsumerManager, error) {
	return &consumerManager{
		consumers:            new(sync.Map),
		consumerGroupClients: new(sync.Map),
	}, nil
}

func (c *consumerManager) GetConsumer(consumerGroup string) (EventMeshConsumer, error) {
	val, ok := c.consumers.Load(consumerGroup)
	if ok {
		return val.(EventMeshConsumer), nil
	}
	cu, err := NewEventMeshConsumer(consumerGroup)
	if err != nil {
		return nil, err
	}
	c.consumers.Store(consumerGroup, cu)
	return cu, nil
}

func (c *consumerManager) RegisterClient(cli *GroupClient) error {
	val, ok := c.consumerGroupClients.Load(cli.ConsumerGroup)
	if !ok {
		cliset := set.New(set.WithGoroutineSafe())
		cliset.Insert(cli)
		c.consumerGroupClients.Store(cli.ConsumerGroup, cliset)
		return nil
	}
	localClients := val.(*set.Set)
	found := false
	for iter := localClients.Begin(); iter.IsValid(); iter.Next() {
		lc := iter.Value().(*GroupClient)
		if lc.GRPCType == consts.WEBHOOK {
			lc.URL = cli.URL
			lc.LastUPTime = cli.LastUPTime
			found = true
			break
		}
		if lc.GRPCType == consts.STREAM {
			lc.Emiter = cli.Emiter
			lc.LastUPTime = cli.LastUPTime
			found = true
			break
		}
	}
	if !found {
		localClients.Insert(cli)
	}
	return nil
}

func (c *consumerManager) DeRegisterClient(cli *GroupClient) error {
	val, ok := c.consumerGroupClients.Load(cli.ConsumerGroup)
	if !ok {
		log.Debugf("no consumer group client found, name:%v", cli.ConsumerGroup)
		return nil
	}
	localClients := val.(*set.Set)
	for iter := localClients.Begin(); iter.IsValid(); iter.Next() {
		lc := iter.Value().(*GroupClient)
		if lc.Topic == cli.Topic {
			if lc.GRPCType == consts.STREAM {
				// TODO
				// close the GRPC client stream before removing it
			}
			localClients.Erase(lc)
		}
	}
	if localClients.Size() == 0 {
		c.consumerGroupClients.Delete(cli.ConsumerGroup)
	}
	return nil
}

func (c *consumerManager) RestartConsumer(consumerGroup string) error {
	val, ok := c.consumers.Load(consumerGroup)
	if !ok {
		return nil
	}
	emconsumer := val.(EventMeshConsumer)
	if emconsumer.ServiceState() == consts.RUNNING {
		if err := emconsumer.Shutdown(); err != nil {
			return err
		}
	}
	if err := emconsumer.Init(); err != nil {
		return err
	}
	if err := emconsumer.Start(); err != nil {
		return err
	}
	if emconsumer.ServiceState() != consts.RUNNING {
		log.Warnf("restart eventmesh consumer failed, status:%v", emconsumer.ServiceState)
		c.consumers.Delete(consumerGroup)
	}
	return nil
}

func (c *consumerManager) UpdateClientTime(cli *GroupClient) {
	val, ok := c.consumerGroupClients.Load(cli.ConsumerGroup)
	if !ok {
		log.Debugf("no consumer group client found, name:%v", cli.ConsumerGroup)
		return
	}
	localClients := val.(*set.Set)
	for iter := localClients.Begin(); iter.IsValid(); iter.Next() {
		iter.Value().(*GroupClient).LastUPTime = time.Now()
	}
}

func (c *consumerManager) clientCheck() {
	sessionExpiredInMills := config2.GlobalConfig().Server.GRPCOption.SessionExpiredInMills
	tk := time.NewTicker(sessionExpiredInMills)
	go func() {
		for range tk.C {
			var consumerGroupRestart []string
			c.consumerGroupClients.Range(func(key, value any) bool {
				localClients := value.(*set.Set)
				for iter := localClients.Begin(); iter.IsValid(); iter.Next() {
					lc := iter.Value().(*GroupClient)
					if time.Now().Sub(lc.LastUPTime) > sessionExpiredInMills {
						log.Warnf("client:%v lastUpdate time:%v over three heartbeat cycles. Removing it",
							lc.ConsumerGroup, lc.LastUPTime)
						emconsumer, err := c.GetConsumer(lc.ConsumerGroup)
						if err != nil {
							log.Warnf("get eventmesh consumer:%v failed, err:%v", lc.ConsumerGroup, err)
							return true
						}
						if err := c.DeRegisterClient(lc); err != nil {
							log.Warnf("deregistry client:%v err:%v", lc.ConsumerGroup, err)
							return true
						}
						if ok := emconsumer.DeRegisterClient(lc); !ok {
							log.Warnf("failed deregistry client:%v in eventmesh consumer", lc.ConsumerGroup)
							return true
						}
						consumerGroupRestart = append(consumerGroupRestart, lc.ConsumerGroup)
					}
				}
				for _, rs := range consumerGroupRestart {
					if err := c.RestartConsumer(rs); err != nil {
						log.Warnf("deregistry consumer:%v  err:%v", rs, err)
						return true
					}
				}
				return true
			})
		}
	}()
}

func (c *consumerManager) Start() error {
	log.Infof("start consumer manager")
	return nil
}

func (c *consumerManager) Stop() error {
	return nil
}
